### DISCUSSION TITLE
[help] Subtle reasons for a target with file dependencies to get outdated on HPC cluster?

### DISCUSSION THREAD

**Discussion by brendanf** [ropensci/targets#1077] [Help]
### Help

- [X] I understand and agree to <https://books.ropensci.org/targets/help.html>.

### Description

I have found that my targets which depend on file targets (i.e., `tar_file()`) tend to become outdated over time.  I specifically notice this on my HPC cluster, which uses LustreFS.  This seems to be somehow non-deterministic; if I check the pipeline after it has just run, no targets are out of date.  But if I check again after 1 month, without touching any files or packages (the pipeline runs inside a container) then some fraction of targets with file dependencies will be outdated.

Yesterday I found that this was the case for one of my pipelines, and I tried to look into it more carefully.  One of the early targets was out of date:
```
> tar_sitrep(protax_refs_full)
# A tibble: 1 × 9
  name             record always never command depend format iteration file
  <chr>            <lgl>  <lgl>  <lgl> <lgl>   <lgl>  <lgl>  <lgl>     <lgl>
1 protax_refs_full FALSE  FALSE  FALSE FALSE   TRUE   FALSE  FALSE     FALSE
```

So then I checked what targets things the dependencies are:
```
> tar_deps({
            outfile <- "data/protax_refs_full.fasta.gz"
            command <- paste(
               "zcat", paste(shQuote(unite_files), collapse = " "), "|",
               "awk", "-f", shQuote(find_unite_seqs_script),
               shQuote(protax_refseq_file), "-", "|",
               "gzip", "-c", "-",
               ">", shQuote(outfile)
            )
            result <- system(command, timeout = 120)
            stopifnot(result == 0)
            outfile
         })
 [1] "<-"                     "=="                     "find_unite_seqs_script"
 [4] "paste"                  "protax_refseq_file"     "shQuote"
 [7] "stopifnot"              "system"                 "unite_files"
[10] "{"
```

Of these, three are previous targets, and the rest are base R. (As a side question, is there a user-facing function to easily get the list of dependencies which targets actually hashes?)

All three of these dependencies are `tar_file()`, some of which are fairly large (>500Mb) and none of the files have changed since the target in question was last run. 

```
> file.info(tar_read(unite_files))$mtime
[1] "2023-01-25 16:33:23 EET" "2023-01-25 16:33:30 EET"
[3] "2023-01-25 16:33:34 EET" "2023-01-25 16:33:40 EET"
[5] "2023-01-25 16:33:51 EET" "2023-01-25 16:33:58 EET"
[7] "2023-01-25 16:34:30 EET"
> file.info(tar_read(protax_refseq_file))$mtime
[1] "2022-08-23 15:43:52 EEST"
> file.info(tar_read(find_unite_seqs_script))$mtime
[1] "2023-01-25 16:10:35 EET"
> tar_meta(protax_refs_full, time)
# A tibble: 1 × 2
  name             time               
  <chr>            <dttm>             
1 protax_refs_full 2023-05-12 21:36:29
```

To add to my confusion, I checked again today and targets no longer considers the target to be outdated.  Fortunately I still had the console open from yesterday when I ran these tests!

Clearly there is something fishy going on here, and I suspect that is is not entirely (or at all!) the fault of targets.  But it would be really great for me to be able to work around it, in a way that still lets the target be outdated when something truly has changed.  I have spent many thousands of CPU hours on re-running targets due to this!

**Comment by wlandau** [comment]
> One of the early targets was out of date:

When this happens, what is the output of `tar_sitrep()` on `protax_refseq_file` and the other files? What does `targets:::file_hash()` return on each, and how does it compare to the `data` column of `tar_meta()`?

> Of these, three are previous targets, and the rest are base R. (As a side question, is there a user-facing function to easily get the list of dependencies which targets actually hashes?)

`dplyr::filter(tar_network(outdated = FALSE)$edges, to == "protax_refs_full")` should show the real dependencies directly upstream.

> All three of these dependencies are tar_file(), some of which are fairly large (>500Mb) and none of the files have changed since the target in question was last run.

It's really strange because the signs point to files having changed, yet the timestamps did not change. 

(NB: one of the changes in development `targets` is that `format = "file"` now ignores timestamps and always rehashes the file, whereas the new `format = "file_fast"` checks the time stamp first to see if the file *could have* changed.)

To try to isolate the issue more closely, maybe you could create a test version of your pipeline with a single (large) file target and prospectively keep an eye on `tar_sitrep()`, `targets:::file_hash()`, `file.info()`, etc.

**Reply by brendanf** [reply]
>>One of the early targets was out of date:

>When this happens, what is the output of tar_sitrep() on protax_refseq_file and the other files? 

`tar_sitrep` showed the other files are not out of date:
```
> tar_sitrep(unite_files)
# A tibble: 1 × 9
  name        record always never command depend format iteration file 
  <chr>       <lgl>  <lgl>  <lgl> <lgl>   <lgl>  <lgl>  <lgl>     <lgl>
1 unite_files FALSE  FALSE  FALSE FALSE   FALSE  FALSE  FALSE     FALSE
> tar_sitrep(find_unite_seqs_script)
# A tibble: 1 × 9
  name                   record always never command depend format itera…¹ file 
  <chr>                  <lgl>  <lgl>  <lgl> <lgl>   <lgl>  <lgl>  <lgl>   <lgl>
1 find_unite_seqs_script FALSE  FALSE  FALSE FALSE   FALSE  FALSE  FALSE   FALSE
# … with abbreviated variable name ¹​iteration
> tar_sitrep(protax_refseq_file)
# A tibble: 1 × 9
  name               record always never command depend format iteration file 
  <chr>              <lgl>  <lgl>  <lgl> <lgl>   <lgl>  <lgl>  <lgl>     <lgl>
1 protax_refseq_file FALSE  FALSE  FALSE FALSE   FALSE  FALSE  FALSE     FALSE
```

>What does targets:::file_hash() return on each, and how does it compare to the data column of tar_meta()?

I didn't check `targets:::file_hash()` from before, but I do still have the results of `tar_meta()` from when I asked this question:
```
> tar_meta(protax_refseq_file)
# A tibble: 1 × 17
  name  type  data  command depend   seed path  time                size   bytes
  <chr> <chr> <chr> <chr>   <chr>   <int> <lis> <dttm>              <chr>  <int>
1 prot… stem  b079… 8acd0b… 01371… 1.27e8 <chr> 2022-08-23 15:43:52 c25d… 1.14e8
# … with 7 more variables: format <chr>, iteration <chr>, parent <chr>,
#   children <list>, seconds <dbl>, warnings <lgl>, error <chr>
# ℹ Use `colnames()` to see all variable names
> tar_meta(find_unite_seqs_script)
# A tibble: 1 × 17
  name  type  data  command depend    seed path  time                size  bytes
  <chr> <chr> <chr> <chr>   <chr>    <int> <lis> <dttm>              <chr> <int>
1 find… stem  0fa6… a9e0fa… ef46d… -9.63e8 <chr> 2023-01-25 16:10:35 7959…  1902
# … with 7 more variables: format <chr>, iteration <chr>, parent <chr>,
#   children <list>, seconds <dbl>, warnings <lgl>, error <chr>
# ℹ Use `colnames()` to see all variable names
> tar_meta(unite_files)
# A tibble: 1 × 17
  name  type  data  command depend   seed path  time                size   bytes
  <chr> <chr> <chr> <chr>   <chr>   <int> <lis> <dttm>              <chr>  <int>
1 unit… stem  0ca8… fa69d0… 88559… 7.11e7 <chr> 2023-01-25 16:34:30 2d77… 5.83e8
# … with 7 more variables: format <chr>, iteration <chr>, parent <chr>,
#   children <list>, seconds <dbl>, warnings <lgl>, error <chr>
# ℹ Use `colnames()` to see all variable names
```

Today, `protax_refs_full` triggered again.  This time I didn't check `tar_outdated()` before starting it, and consequently didn't notice that it was re-running the target before it finished, but here are the current results for its prerequisites.

```
> tar_meta(c(protax_refseq_file, find_unite_seqs_script, unite_files))
# A tibble: 3 × 17
  name        type  data  command depend    seed path  time                size 
  <chr>       <chr> <chr> <chr>   <chr>    <int> <lis> <dttm>              <chr>
1 protax_ref… stem  b079… 8acd0b… 01371…  1.27e8 <chr> 2022-08-23 15:43:52 c25d…
2 find_unite… stem  0fa6… a9e0fa… ef46d… -9.63e8 <chr> 2023-01-25 16:10:35 7959…
3 unite_files stem  0ca8… fa69d0… 88559…  7.11e7 <chr> 2023-01-25 16:34:30 2d77…
# … with 8 more variables: bytes <int>, format <chr>, iteration <chr>,
#   parent <chr>, children <list>, seconds <dbl>, warnings <lgl>, error <chr>
# ℹ Use `colnames()` to see all variable names
> targets:::file_hash(tar_read(protax_refseq_file))
[1] "b07971fb4e17ff0f"
> targets:::file_hash(tar_read(find_unite_seqs_script))
[1] "0fa68a325730a9f2"
> targets:::file_hash(tar_read(unite_files))
[1] "0ca81a7bce15b614"
```

There don't seem to be any changes.

>(NB: one of the changes in development targets is that format = "file" now ignores timestamps and always rehashes the file, whereas the new format = "file_fast" checks the time stamp first to see if the file could have changed.)

I will definitely try that out.

**Reply by wlandau** [reply]
> tar_sitrep showed the other files are not out of date:

So it looks like one of the other (non-file) dependencies must be outdated. I wonder if it could be isolated to a function or global object, e.g. with `tar_network()$edges`. For example, `R6` classes and ad hoc environments have unpredictable internals which make hashes change a lot.

**Comment by wlandau** [comment]
If all else fails, you can set `cue = tar_cue(file = FALSE)` for the `cue` argument of each file target in order to ignore spurious changes to the hashes of those files.
